### PR TITLE
gaudi: disentangle variant 'optional'

### PR DESCRIPTION
**Pull request by wdconinc**
The catch-all `optional` variant included a few not-quite-so-optional components. This attempts to split the `optional` variant into a number of more fine-grained variants.

TODO:
- [x] maintain backwards-compatibility by defining `optional` as equivalent to the corresponding set of new variants.

Maintainers: @vvolkl @drbenmorgan

### CODE REVIEW

**Review by vvolkl — DISMISSED** [review]
Makes a lot of sense to me, thanks!

**Comment by wdconinc** [comment]
`optional` is implemented as a meta variant that conflicts with `-{v}` for `v` in the list of new variants. The concretizer then figures out to set these variants all too `+{v}`.

Note: The following now will throw an obvious conflict: `spack spec gaudi +optional -aida`, so you cannot use the `+optional` variant as a starting point from which to 'subtract' functionality.

File: var/spack/repos/builtin/packages/gaudi/package.py [hunk 1]
```diff
@@ -57,40 +85,33 @@ class Gaudi(CMakePackage):
     depends_on("cmake", type="build")
     depends_on("cppgsl")
     depends_on("fmt", when="@33.2:")
-    depends_on("intel-tbb")
+    depends_on("tbb")
```
Comment (wdconinc) [inline var/spack/repos/builtin/packages/gaudi/package.py]: Leave `intel-tbb` dependency unchanged, but probably needs to be addressed in a separate PR when we can verify that gaudi builds fine with the other `tbb` providers, `intel-oneapi-tbb` and `intel-parallel-studio`.
```suggestion
    depends_on("intel-tbb")
```

File: var/spack/repos/builtin/packages/gaudi/package.py [hunk 2]
```diff
@@ -63,34 +91,27 @@ class Gaudi(CMakePackage):
...
-    depends_on("py-setuptools", when="+optional")
-    depends_on("py-nose", when="+optional")
-    depends_on("relax", when="@:33 +optional")
-    depends_on("xerces-c", when="+optional")
+    depends_on("gperftools", when="+gperftools")
+    depends_on("gdb")
+    depends_on("heppdt", when="+heppdt")
+    depends_on("jemalloc", when="+jemalloc")
+    depends_on("libunwind", when="+unwind")
+    # depends_on("relax", when="@:33 +optional")
```
Comment (wdconinc) [inline var/spack/repos/builtin/packages/gaudi/package.py]: RELAX is required through v34r1 for examples, https://gitlab.cern.ch/gaudi/Gaudi/-/blob/v34r1/GaudiExamples/CMakeLists.txt#L26.
```suggestion
    depends_on("relax", when="@:34 +examples")
```

File: var/spack/repos/builtin/packages/gaudi/package.py [hunk 3]
```diff
@@ -63,34 +91,27 @@ class Gaudi(CMakePackage):
...
-    depends_on("py-networkx", when="+optional ^python@3.0.0:")
-    depends_on("py-setuptools", when="+optional")
-    depends_on("py-nose", when="+optional")
-    depends_on("relax", when="@:33 +optional")
-    depends_on("xerces-c", when="+optional")
+    depends_on("gperftools", when="+gperftools")
+    depends_on("gdb")
+    depends_on("heppdt", when="+heppdt")
+    depends_on("jemalloc", when="+jemalloc")
+    depends_on("libunwind", when="+unwind")
```
Comment (wdconinc) [inline var/spack/repos/builtin/packages/gaudi/package.py:94]: `libpng` is required through v34r1 for examples, https://gitlab.cern.ch/gaudi/Gaudi/-/blob/v34r1/GaudiExamples/CMakeLists.txt#L29.
```suggestion
    depends_on("libng", when="@:34 +examples")
    depends_on("libunwind", when="+unwind")
```

File: var/spack/repos/builtin/packages/gaudi/package.py [hunk 4]
```diff
@@ -63,34 +91,27 @@ class Gaudi(CMakePackage):
...
-    depends_on("libunwind", when="+optional")
-    depends_on("py-networkx@:2.2", when="+optional ^python@:2.7")
-    depends_on("py-networkx", when="+optional ^python@3.0.0:")
-    depends_on("py-setuptools", when="+optional")
-    depends_on("py-nose", when="+optional")
-    depends_on("relax", when="@:33 +optional")
-    depends_on("xerces-c", when="+optional")
+    depends_on("gperftools", when="+gperftools")
+    depends_on("gdb")
+    depends_on("heppdt", when="+heppdt")
```
Comment (wdconinc) [inline var/spack/repos/builtin/packages/gaudi/package.py:91]: `HepPDT` is required through v34r1 for examples, https://gitlab.cern.ch/gaudi/Gaudi/-/blob/v34r1/GaudiExamples/CMakeLists.txt#L16. After that, `HepPDT is optional in `PartPropSvc`.
```suggestion
    depends_on("heppdt", when="@:34 +examples")
    depends_on("heppdt", when="+heppdt")
```

File: var/spack/repos/builtin/packages/gaudi/package.py [hunk 5]
```diff
@@ -63,34 +91,27 @@ class Gaudi(CMakePackage):
...
-    depends_on("libpng", when="+optional")
-    depends_on("libunwind", when="+optional")
-    depends_on("py-networkx@:2.2", when="+optional ^python@:2.7")
-    depends_on("py-networkx", when="+optional ^python@3.0.0:")
-    depends_on("py-setuptools", when="+optional")
-    depends_on("py-nose", when="+optional")
-    depends_on("relax", when="@:33 +optional")
-    depends_on("xerces-c", when="+optional")
+    depends_on("gperftools", when="+gperftools")
+    depends_on("gdb")
```
Comment (wdconinc) [inline var/spack/repos/builtin/packages/gaudi/package.py:88]: `GSL`  is required through v31r0 for examples, https://gitlab.cern.ch/gaudi/Gaudi/-/blob/v31r0/GaudiExamples/CMakeLists.txt#L10.
```suggestion
    depends_on("gdb")
    depends_on("gsl", when="@:31 +examples")
```

File: var/spack/repos/builtin/packages/gaudi/package.py [hunk 6]
```diff
@@ -35,11 +35,39 @@ class Gaudi(CMakePackage):
...
+    ]
+    variant(
+        "optional",
+        default=False,
+        description="Build optional components; equivalent to{}".format(
+            " +".join([""] + optional_variants)
+        ),
+    )
+    for v in optional_variants:
+        conflicts("-" + v, when="+optional")
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/gaudi/package.py]: Hmm, variant in other packages change frequently, so I wouldn't write these 20 lines but just change the variants.

Comment (wdconinc) [reply var/spack/repos/builtin/packages/gaudi/package.py]: @vvolkl Dropping `optional` ok with you?

Comment (vvolkl) [reply var/spack/repos/builtin/packages/gaudi/package.py]: ok with me!

Comment (wdconinc) [reply var/spack/repos/builtin/packages/gaudi/package.py]: Then we leave this as something local that key4hep or eic spack repos can add in an adhoc way as an overlay.

```suggestion
```

**Review by alalazo — APPROVED** [review]
